blog/_posts/2020-11-24-ranluxpp.md

```diff
@@ -0,0 +1,157 @@
+[By adding `#pragma` directives][unrolling], it is possible to help the compiler.
+As can be seen in above graph, this improves performance for GCC 8.3.1 and Clang 9.0.1 by up to 30%.
+GCC 7.5.0 does not benefit from the change because `#pragma GCC unroll` was only added in GCC 8.
+
+[unrolling]: https://github.com/root-project/root/commit/0f0a8077a9e9902e08ff0f8cc93644f00ef33559
+
+### Overflow handling
+
+For the multiplication of two 576 bit numbers, the portable code is constrained by the available data types:
+it can at most multiply two 32 bit values at a time to stay within the 64 bits of `uint64_t`.
+This makes it necessary to handle overflows when summing up the partial results.
+[By arranging the operations differently][overflow], it is possible to avoid conditional execution.
+Together with full unrolling, this eliminates all jumps from the generated machine code for GCC 8.3.1.
+During execution, this leads to performance improvements of up to 15% for the two versions of GCC.
+There is no change for Clang because its optimizations already transformed the original code.
+
+[overflow]: https://github.com/root-project/root/commit/2856d170525c29b6dffe220387a0a8e3671a2e7a
+
+### Native multiplication
+
+An even faster way is to let hardware handle the multiplication of two 64 bit numbers.
+This is possible with the extension type `__int128` supported by both GCC and Clang.
+Using this type, it is possible to multiply 64 bit numbers with 128 bit precision.
+For that, the compiler can make use of special instructions available for the targeted hardware.
+Afterwards the upper and lower halves can be extracted as 64-bit numbers.
+The [change taking advantage of `__int128`][native] improves performance of all three compilers by more than 40%!
+
+[native]: https://github.com/root-project/root/commit/e37bfe378c50e5d1f638a44801843123c678ce19
+
+### Carry propagation
+
+Additional tests showed inferior performance with older compilers.
+Part of this is related to missing loop unrolling because the `#pragma` is not available (see above).
+An additional problem is the generation of jumping code to propagate carry bits in case of overflows.
+Instead of relying on inline assembly code, I found a [portable solution][carry] after [refactoring the common code][prep].
+The modification results in good code for all tested compilers.
+For the older GCC 7.5.0 in particular, it improves performance by more than 20%.
+
+[prep]: https://github.com/root-project/root/commit/f081d391e6b05d85520701b940c008107006313c
+[carry]: https://github.com/root-project/root/commit/28dd98b3f7d4a3a43a331b9f6483fcdc19244791
+
+## Conclusion
+
+In this blog post, I discussed the implementation of RANLUX++ and its performance.
+After tuning, the average time per number is down to around 9ns, very close to the assembly version.
+It is slightly slower than `TRandom3` currently used for `gRandom` (around 3ns per number).
```

Comment:
I would also add here that TRandom3 is only a 32 bits generators while new RANLUX++ is a 52 bits